### PR TITLE
!!! FEATURE: Overhaul catch up implementation

### LINKED ISSUE #4289: Overhaul CatchUp and CatchUpHook implementation
Currently there's a lot of duplication in the `catchUp()` methods of projections and support for catch up hooks has to be implemented manually (see https://github.com/neos/neos-development-collection/blob/448a2d151a0b31a68882e7c86a583f5845fe3133/Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php#L187-L197 for example).

We should probably replace
```php
public function catchUp(EventStreamInterface $eventStream, ContentRepository $contentRepository): void
```

by something like

```php
public function apply(EventInterface $event): void
```

in the `ProjectionInterface` and move the looping/catchup hook logic out.

Projections that need more control over the whole event stream could opt in with some additional interface *if we need it* at all.

### PR DESCRIPTION
**Pull request by bwaidelich**
Mainly adjusts the [ProjectionInterface](https://github.com/neos/neos-development-collection/pull/4405/files#diff-8af18a3ef17c7e94b4ca89b55ad24d2e43c14a0163703d98f5cb4282ef3ce783) such that the iteration of events can happen outside of the implementation.

This allows to centralize logic which enables us to:
* Support CatchUp Hooks for all core and 3rd party projections
* Improve error handling (#3854)
* Introduce a progress bar for projection replay/catchup on the CLI (#4200)
* Simplify projection implementation because event denormalization happens in the outer layer
* In general: Give us more control over catch ups

## Breaking Change

This is a breaking change only if you implemented a custom implementation of the `ProjectionInterface`:

### 1. `canHandle()` now gets an instance of the domain `EventInterface` instead of the low level ES event

The code can be adjusted like this:

```php
    public function canHandle(Event $event): bool
    {
        $eventClassName = $this->eventNormalizer->getEventClassName($event);
        // ...
```

=> 

```php
    public function canHandle(EventInterface $event): bool
    {
        $eventClassName = $event::class;
        // ...
```

### 2. `catchUp()` has been replaced with `apply()`

The code can be adjusted like this:

```php
    public function catchUp(EventStreamInterface $eventStream, ContentRepository $contentRepository): void
    {
        // any catch up hook initialization here...
        $catchUp = CatchUp::create($this->apply(...), $this->checkpointStorage);
        $catchUp->run($eventStream);
    }

    private function apply(EventEnvelope $eventEnvelope): void
    {
        if (!$this->canHandle($eventEnvelope->event)) {
            return;
        }

        $event = $this->eventNormalizer->denormalize($eventEnvelope->event);
        // projection logic here...
```

=> 

```php
    public function apply(EventInterface $event): void
    {
        // projection logic here...
```

*Note:* The `EventEnvelope` gets passed in as 2nd argument and can be implemented if access to low-level details (e.g. event metadata) is needed:

```php
    public function apply(EventInterface $event, EventEnvelope $eventEnvelope): void
    {
        // projection logic here...
```

### 3. `getSequenceNumber()` was replaced by `getCheckpointStorage()`

This enables the framework to use the `CatchUp` helper in the outer layers and it will allow us to implement better "blocking" mechanisms in the future

The code can be adjusted like this:

```php
    public function getSequenceNumber(): SequenceNumber
    {
        return $this->checkpointStorage->getHighestAppliedSequenceNumber();
    }
```

=>

```php
    public function getCheckpointStorage(): CheckpointStorageInterface
    {
        return $this->checkpointStorage;
    }
```

Resolves: #4289

### CODE REVIEW

**Comment by skurfuerst** [comment]
@bwaidelich IMHO this looks really good. Tests work still (as expected <3) So I'd say this is pretty much merge ready? (Not merging yet because the checkboxes ar enot yet ticked ;) )

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php [hunk 1]
```diff
@@ -44,25 +44,23 @@
...
+use Neos\EventStore\CatchUp\CheckpointStorageInterface;
 use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
-use Neos\EventStore\Model\Event;
+use Neos\EventStore\Model\Event\SequenceNumber;
 use Neos\EventStore\Model\EventEnvelope;
 use Neos\EventStore\Model\EventStore\SetupResult;
-use Neos\EventStore\Model\EventStream\EventStreamInterface;
-use Neos\EventStore\Model\Event\SequenceNumber;
+
+use function get_debug_type;
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php]: i wasnt sure what we discussed, if we now like to adapt this code style or not - but it turns out we have yet no 100% clear decision: https://neos-project.slack.com/archives/C04Q0TC15/p1692352868964289?thread_ts=1683558218.323329&cid=C04Q0TC15

File: Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/HypergraphProjection.php [hunk 2]
```diff
@@ -45,17 +44,15 @@
...
+use Neos\ContentRepository\Core\Projection\ProjectionInterface;
+use Neos\EventStore\CatchUp\CheckpointStorageInterface;
 use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
-use Neos\EventStore\Model\Event;
 use Neos\EventStore\Model\Event\SequenceNumber;
 use Neos\EventStore\Model\EventEnvelope;
 use Neos\EventStore\Model\EventStore\SetupResult;
-use Neos\EventStore\Model\EventStream\EventStreamInterface;
+
+use function sprintf;
```
Comment (dlubitz) [inline Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/HypergraphProjection.php]: There is another use function, I guess by accident.

```suggestion
```

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 3]
```diff
@@ -121,26 +130,64 @@ public function handle(CommandInterface $command): CommandResult
...
+            return $projectionState;
+        }
+        foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+            $projectionState = $projection->getState();
+            if ($projectionState instanceof $projectionStateClassName) {
+                $this->projectionStateCache[$projectionStateClassName] = $projectionState;
+                return $projectionState;
+            }
+        }
+        throw new \InvalidArgumentException(sprintf('a projection state of type "%s" is not registered in this content repository instance.', $projectionStateClassName), 1662033650);
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]:
```suggestion
        throw new \InvalidArgumentException(sprintf('A projection state of type "%s" is not registered in this content repository instance.', $projectionStateClassName), 1662033650);
```

File: Neos.ContentRepository.Core/Classes/Projection/CatchUpHookFactories.php [hunk 4]
```diff
@@ -46,8 +46,7 @@ private function has(string $catchUpHookFactoryClassName): bool
 
     public function build(ContentRepository $contentRepository): CatchUpHookInterface
     {
-        $catchUpHooks = array_map(fn(CatchUpHookFactoryInterface $catchUpHookFactory) =>
-            $catchUpHookFactory->build($contentRepository), $this->catchUpHookFactories);
+        $catchUpHooks = array_map(static fn(CatchUpHookFactoryInterface $catchUpHookFactory) => $catchUpHookFactory->build($contentRepository), $this->catchUpHookFactories);
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Projection/CatchUpHookFactories.php:49]: Ohh boy, I'm still learning new things 🙈 

👌

File: Neos.ContentRepository.Core/Classes/Projection/ContentStream/ContentStreamProjection.php [hunk 5]
```diff
@@ -19,31 +19,30 @@
...
+use Neos\ContentRepository\Core\Projection\ProjectionStateInterface;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\EventStore\CatchUp\CheckpointStorageInterface;
 use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
-use Neos\EventStore\Model\Event;
 use Neos\EventStore\Model\Event\SequenceNumber;
 use Neos\EventStore\Model\EventEnvelope;
-use Neos\EventStore\Model\EventStream\EventStreamInterface;
+
+use function sprintf;
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Projection/ContentStream/ContentStreamProjection.php]:
```suggestion
```

File: Neos.ContentRepository.Core/Classes/Projection/NodeHiddenState/NodeHiddenStateProjection.php [hunk 6]
```diff
@@ -19,21 +19,19 @@
...
-use Neos\ContentRepository\Core\Projection\ProjectionStateInterface;
-use Neos\EventStore\CatchUp\CatchUp;
+use Neos\EventStore\CatchUp\CheckpointStorageInterface;
 use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
-use Neos\EventStore\Model\Event;
 use Neos\EventStore\Model\Event\SequenceNumber;
 use Neos\EventStore\Model\EventEnvelope;
-use Neos\EventStore\Model\EventStream\EventStreamInterface;
+
+use function sprintf;
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Projection/NodeHiddenState/NodeHiddenStateProjection.php]:
```suggestion
```

File: Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceProjection.php [hunk 7]
```diff
@@ -19,31 +19,30 @@
...
 use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
-use Neos\EventStore\CatchUp\CatchUp;
+use Neos\EventStore\CatchUp\CheckpointStorageInterface;
 use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
-use Neos\EventStore\Model\Event;
 use Neos\EventStore\Model\Event\SequenceNumber;
 use Neos\EventStore\Model\EventEnvelope;
-use Neos\EventStore\Model\EventStream\EventStreamInterface;
+
+use function sprintf;
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceProjection.php]:
```suggestion
```

File: Neos.Neos/Classes/AssetUsage/Projection/AssetUsageProjection.php [hunk 8]
```diff
@@ -4,40 +4,38 @@
...
 use Neos\Media\Domain\Repository\AssetRepository;
-use Neos\Media\Domain\Model\AssetVariantInterface;
-use Neos\Media\Domain\Model\AssetInterface;
 use Neos\Neos\AssetUsage\Dto\AssetIdAndOriginalAssetId;
+use Neos\Neos\AssetUsage\Dto\AssetIdsByProperty;
+use Neos\Neos\AssetUsage\Dto\NodeAddress;
+use Neos\Utility\Exception\InvalidTypeException;
+use Neos\Utility\TypeHandling;
+
+use function sprintf;
```
Comment (dlubitz) [inline Neos.Neos/Classes/AssetUsage/Projection/AssetUsageProjection.php]:
```suggestion
```

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php (hunk 1 above)
Comment (dlubitz) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php]:
```suggestion
```

**Review by dlubitz — APPROVED** [review]
This looks great. I just left some nitpic comments. 👍

File: Neos.ContentRepository.Core/Classes/CommandHandler/PendingProjections.php [hunk 9]
```diff
@@ -96,20 +97,23 @@ public static function fromProjectionsAndEventsAndSequenceNumber(
         SequenceNumber $highestCommittedSequenceNumber
     ): self {
         $sequenceNumberInteger = $highestCommittedSequenceNumber->value - $events->count() + 1;
-        $pendingProjections = Projections::create();
+        $pendingProjectionsArray = [];
         $sequenceNumberPerProjection = [];
         foreach ($events as $event) {
             foreach ($allProjections as $projection) {
+                if ($event instanceof DecoratedEvent) {
```
Comment (nezaniel) [inline Neos.ContentRepository.Core/Classes/CommandHandler/PendingProjections.php]: This can be moved to the outer loop, can't it?

**Comment by nezaniel** [comment]
Otherwise, looks great! Let's merge

File: Neos.ContentRepository.Core/Classes/CommandHandler/PendingProjections.php (hunk 9 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/CommandHandler/PendingProjections.php]: True, good catch

**Review by nezaniel — APPROVED** [review]
Looks nice, let's do this!

**Pull request merged**